Upload session

iget-master opened the pull request:
This PR proposes a fix (actually a feature) to #10.

Since Dropbox API supports uploads up to 150MB in a single request, we should break it into many chunks and use the `upload_session` api feature.

This PR implement the following public methods to the Client:

- uploadSessionStart
- uploadSessionAppend
- uploadSessionFinish
- uploadChunked

The last one is the main method that should be used outside the client, but I made the other methods public if someone want to do it own implementation (someone may need split the uploads in many processes to speedup or distribute load).

**Please don't merge this pull request until I check this points:**

- [x] Add PHPDOC to new methods
- [x] Test on my local environment for large files upload

freekmurze commented on src/Client.php:
```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+            // Start upload session on first iteration, then just append on subsequent iterations
+            if (isset($cursor)) {
+                $cursor = $this->uploadSessionAppend($data, $cursor);
+            } else {
```
I really dislike `else` statements. Could you replace this by using a ternary operator?

freekmurze commented on src/Client.php:
```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+            $stream = fopen('php://memory', 'r+');
+            fwrite($stream, $contents);
+            rewind($stream);
+        } else {
```
Could you rewrite this in a way that an `else` statement is not needed?

freekmurze commented on src/Client.php:
```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+     * @throws Exception
+     * @return string
+     */
+    private static function readFully($inStream, int $numBytes)
```
Make this `protected`

freekmurze commented on src/Client.php:
```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+     */
+    private static function readFully($inStream, int $numBytes)
+    {
+        $full = '';
```
rename this to `fileContents`

iget-master replied:
This method was copied from original `dropbox-sdk-php`.

I think that makes more sense calling it `readChunk` since it gets a chunk from the stream.

freekmurze commented on src/Client.php:
```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+    private static function readFully($inStream, int $numBytes)
+    {
+        $full = '';
+        $bytesRemaining = $numBytes;
```
Generally speaking I don't like abbreviations. Rename `numBytes` to `numberOfBytes`

iget-master replied:
I agree, I'm against abbreviations. I'll fix it.

freekmurze commented on src/UploadSessionCursor.php:
```diff
@@ -0,0 +1,32 @@
+     *
+     * @var string
+     */
+    public $session_id;
```
Rename all instances of `session_id` to `sessionId`

iget-master replied:
Renaming to to lowerCamelCase have a downside: I will need to implement a `toArray()` method, since on current way, the object is directly converted to array by guzzlehttp. Are you sure?

freekmurze replied:
Didn't know that.

Ok, leave it like it is now.

freekmurze commented on src/UploadSessionCursor.php:
```diff
@@ -0,0 +1,32 @@
+    public $offset;
+
+    /**
+     * UploadSessionCursor constructor.
```
Remove this redundant docblock

freekmurze commented:
Could you also test in the current `upload` method if the given file is more than 150 MB in size? If it is just leverage your new `uploadChunked` method.

In that way user of this package don't have to think about which method to use. The package will just do the right thing.

iget-master commented:
> Could you also test in the current upload method if the given file is more than 150 MB in size? If it is just leverage your new uploadChunked method.
> 
> In that way user of this package don't have to think about which method to use. The package will just do the right thing.

Actually there a problem with this. Some streams cannot have it size read, eg: `popen("mysqldump ...")`.

To read it size, I think that I will need to load it all on memory to measure.
But I'll look how the old SDK does it to try to replicate.

iget-master commented:
At [old sdk](https://github.com/dropbox/dropbox-sdk-php/blob/master/lib/Dropbox/Client.php#L274-L297), that's the behavior: 

If the user pass the file size as parameter to `uploadFile` method, it check the necessity of chunking. Otherwise, it will chunk anyway.

iget-master commented:
I've tested the current solution to upload many combinations of files: 

- 1GB split in 128M chunks (session start, append(many times) and finish used)
- 2KB split in 1KB chunks (only session start and session finish is used)
- 100B split in 1M chunks (only session start will send data)

iget-master commented:
@freekmurze Please take a look at this implementation: https://github.com/spatie/dropbox-api/pull/21/commits/a444d41c74a6e2b44d89ac619b3831740d62109c

It tries to figure out if it's necessary to use chunked upload method or not.

iget-master commented:
@freekmurze I've implemented the `shouldUploadChunked` method as you asked.

Everything is OK, except for coverage. 

I'm looking a way to test this line: https://github.com/iget-master/dropbox-api/blob/4d7ba3944dbe7148a2e330207df8eea9f016dc14/src/Client.php#L461

freekmurze commented:
@iget-master I don't care much for 100% code coverage. I'm ok with not having a test for that code if it is too difficult to set up such a test.

I'll do  a review of this PR when I've got some time on my hands.

iget-master commented:
Okay. I agree about the coverage. To cover that line I'll need to mock the filesystem just to prove that a simple if statement is workink! :-1: 

Thank you!

iget-master commented:
@freekmurze the https://github.com/spatie/dropbox-api/pull/21/commits/c1d286cecce8285cfc35b9bfff779c483ced3f85 commit is about fixing/workaround a issue related to https://github.com/guzzle/psr7/issues/155

freekmurze commented:
@iget-master That workaround code is pretty grim. I'll wait a few days with pulling this in hoping that problem will be fixed in Guzzle itself.

iget-master commented:
@freekmurze I noticed that even if guzzle fixes the issue, we'll still need the `isPipe` check. 

Also noticed that guzzle doesn't respond to any PR or issues for months! :-1:

freekmurze commented on src/Client.php:
```diff
@@ -20,6 +20,8 @@ class Client
     const THUMBNAIL_SIZE_L = 'w640h480';
     const THUMBNAIL_SIZE_XL = 'w1024h768';
 
+    const MAX_CHUNK_SIZE = 157286400;
```
It's very difficult to know how many megabytes this number represents

Could you rewrite this as `1024 * 1024 * <amount of megabytes>`?

iget-master replied:
Done

freekmurze commented:
Are there any other things you'd like to change before I merge this awesome contribution?

Have you tested it in a real app that this all works?

iget-master commented:
@freekmurze I've tested only on simulated application, sending our backup to dropbox using it, but seems to work fine, I've tried some different file sizes and situations, everything OK.

I think that this can be merged out.

chris-faulkner commented:
I have this package set up on a live server and would be happy to test it for you. I'll certainly be testing it as soon as you merge it in! Really appreciate the work you guys do and hope to contribute in future when I'm a bit more experienced.

freekmurze commented:
@chris-faulkner testing out some changes is a great way to contribute/help us as well.

I'll ping you as soon as I release this changes.

freekmurze commented:
`v1.6.0` has been tagged

@chris-faulkner can you test if this works for you?

chris-faulkner commented:
@freekmurze I just tested it and it seemed to be working but then gave me the following [error](https://pastebin.com/raw/QLrhH4t6)
I've tried to figure out what's wrong and it may not be related to this but I have to go out for the day so I'll have another look when I can.

chris-faulkner commented:
So it turns out the files copied fine, not sure why it's giving the error?

freekmurze commented:
@iget-master could you provide a fix for @chris-faulkner 's issue?

iget-master commented:
@freekmurze I just did a new commit on my fork to fix the issue. See https://github.com/iget-master/dropbox-api/commit/2b946a34c81e0d83df3006b02dd3670a49054b51

The issue may not be an issue caused by dropbox-api, but I've changed things to still working:

The flysystem-dropbox relies on a `.tag` attribute that's not present on the dropbox api. I've added it as it's done on other pieces of code.

https://github.com/spatie/flysystem-dropbox/blob/master/src/DropboxAdapter.php#L292

iget-master commented:
See #24

arnoldjp57 commented:
Hi... A Quick question .... I have done a composer update and I get the dropbox-api version 1.6.0 ... If I check, the latest version is 1.6.1 ....Any Idea my composer not "seeing" the latest version? I have manually changed the Client.php and my dropbox uploads are working like a charm
Please Advise .... Regards from South Africa :-)

iget-master commented:
@arnoldjp57 looks like packagist (composer's repository) didn't noticed the new version. Maybe @freekmurze need to fix it manually.

freekmurze commented:
Let me check...

freekmurze commented:
Fixed it, should be better now.

Run `composer update` again.

arnoldjp57 commented:
Hi ... Fixed ... Thank you very much
